#!/bin/bash

# Example usage via condor submission system on hydrogen node7
# csmit -m 10G -c 1 "bash alignment_stats.sh WT_RNAseq_Chris_Rep1"

i=$1

### Calculate alignment summary statistics for fastq files and for each bam file generated by the pipeline

# Output files with the suffix "fastq.stats" or "bam.stats" contain read-pair counts,
# while output files with the suffix "alignments.stats" contain alignment counts
# This is to provide distinct alignment counts where reads map to multiple loci
gunzip -c -k ${i}_R1.fastq.gz > ${i}_R1.fastq
gunzip -c -k ${i}_R2.fastq.gz > ${i}_R2.fastq
cat ${i}_R1.fastq | echo $((`wc -l`/4)) > ${i}_R1.fastq.stats
cat ${i}_R2.fastq | echo $((`wc -l`/4)) > ${i}_R2.fastq.stats

paste -d "\t" ${i}_R1.fastq.stats ${i}_STAR_mapped.bam.stats ${i}_STAR_mapped_unique_sort.bam.stats ${i}_STAR_mapped_multi.bam.alignments.stats ${i}_STAR_mapped_multi_primary_alignment.bam.alignments.stats ${i}_STAR_mapped_multi_primary_alignment.bam.stats ${i}_STAR_mapped_multi_primary_alignment_fq3_sort.bam.stats ${i}_STAR_mapped_both_sort.bam.stats > ${i}_alignment_summary_stats.txt
sed -i '1i Total_sequenced_read_pairs\tAligned,_mismatches<=2\tUniquely_aligned\tMultiply_aligned_read_pair_alignments\tPrimary_multiply_aligned_read_pair_alignments\tPrimary_multiply_aligned\tPrimary_multiply_aligned,_MAPQ==3\tBoth' ${i}_alignment_summary_stats.txt
[ -d alignment_stats ] || mkdir alignment_stats
mv ${i}*.stats alignment_stats/
mv ${i}_alignment_summary_stats.txt alignment_stats/


### Remove no longer required fastq files
rm ${i}_R1.fastq ${i}_R2.fastq

